### DISCUSSION TITLE
Consider minor mode to reduce user configuration boilerplate

### DISCUSSION THREAD

**Discussion by jeffvalk** [minad/consult#1054] [Ideas]
Many Consult users will prefer Consult's functions to built-in equivalents wherever available. Currently, this must be configured one function at a time, which produces a lengthy configuration (as shown in the `use-package` example in the README). A concise way to enable Consult's improvements to built-in functions would be a welcome enhancement and feel more "batteries included".

One possible approach would be to define a minor mode in which relevant built-in functions are remapped to the Consult equivalents. Enabling the minor mode wouldn't be required, and disabling the mode would disable the remapping. There would be no loss of flexibility, only an additional DWIM way to enable Consult. (The Counsel package does something [along these lines][1].) Might this be worth considering?

[1]: https://github.com/abo-abo/swiper/blob/2a25a6fb5b081cb141c5eccac8ee58ab1feeb747/counsel.el#L7080-L7129

**Comment by minad** [comment]
Such a mode intentionally hasn't been part of Consult for multiple reasons:

- Consult aims to stay out of the way and not override functionality out of the box.
- Users may only want to use a subset of the functionality.
- For many commands there is no canonical key binding.
- It doesn't feel right if the whole purpose of the minor mode is to override key bindings, scattered across many key maps. One could rather provide a prefix map which could be bound to a Consult prefix key, but this wouldn't yield the same convenience.
- If you want a preconfigured setup, you can consider trying one of the many Emacs distributions.
- ...

**Comment by jeffvalk** [comment]
I fully agree that overriding key bindings by default would be bad behavior. I'm thinking about how to make user customization expressive of the user's intention, rather than requiring the user to enumerate more than is necessary.

Consult's design includes knowledge of which Consult functions replace which built-in functions. (The `use-package` example in the README suggests that there are 15 of these.) Why not expose this information to the user in the code? If not a minor mode, a simple alist variable could improve both discoverability and configuration succinctness.

**Reply by minad** [reply]
I see your point that it would be nice to have a simplified setup. This applies to all my packages. For example just yesterday I've added a lazily loaded `cape-prefix-map` to [Cape](https://github.com/minad/cape/commit/040c1ee30b6ff766ba0f3e1458b43ce4a65f068b), which allows to conveniently access all completion commands conveniently under a common prefix key. However the situation in Consult is much more heterogeneous. There we don't install a single map but override bindings across many maps. A global minor mode with the only purpose of installing some configurations is not an ideal solution, I even consider it a malpractice. In contrast to the `cape-prefix-map`, such a minor mode necessarily has to be loaded and cannot be activated lazily.
